Add update mechanism for :penguin:

stanier opened the pull request:
Temporary(?) solution to #2956 while better solution is discussed.

stanier committed:
Added update mechanism for :penguin

stanier committed:
Tying up loose ends from last commit

p-e-w commented:
This PR is a good shot at fixing the single biggest issue on Linux. It has now been tagged "needs-review" for almost a month. Any chance of speeding up the review process so there can be progress on this issue?

The way I see it, while the approach of trying to identify the distro and then downloading packages is not pretty, it's the only way to move forward given the circumstances. There is one – and only one – proper way to install and update packages on Linux: From repositories. If the Atom team is unwilling to create official aptitude/yum repos, the "semi-manual" approach is the next best thing. Using third-party repos is simply not an option in many environments due to security concerns.

antond commented:
This would be great to have. I hope it gets merged soon.

Zireael07 commented:
Has it been reviewed at all? There are no comments to this effect...

stanier commented:
@Zireael07 I can't really attest to the functionality of this fix as it was only tested on two local builds and I didn't really care to mess with CoffeeScript's syntactic bull.  I think the only part that really needs "reviewing" is the decision of whether or not to continue using the release channel for updates or to make use of a ppa (as we await a response in #2956).

In my personal opinion, the update process should be automated regardless.  I might push a commit later today to touch up some things and make it more likely to be implemented.

stanier committed:
Added sugar to the :coffee:

Added in the CoffeeScript syntactic sugar to make Travis stop complaining.

stanier committed:
Replaced if statements with else if

I was tired, OK?

Also removed my stupid comments.  I like to think I'm a funny person.  I'm not.

50Wliu commented on src/browser/auto-update-manager.coffee:
```diff
@@ -41,8 +43,8 @@ class AutoUpdateManager
       @setState(CheckingState)
-
     autoUpdater.on 'update-not-available', =>
+
```
Why was this line added, and why was the above one removed?

50Wliu commented on src/browser/auto-update-manager.coffee:
```diff
@@ -59,7 +61,7 @@ class AutoUpdateManager
       when 'win32'
```
This can be combined into `when 'win32', 'linux'`

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+
+  stdout = ChildProcess.execSync('cat /etc/issue').toString().replace(/\\./g, "")
+
+  if distros.ubuntu.test(stdout) | distros.debian.test(stdout)
```
Shouldn't this be using `||` instead of `|`?

stanier committed:
Fixed line spacing

Caused by accidental line swap that went unnoticed.

Apparently I have a bad habit of pressing the control key without realizing it.

stanier committed:
Combined switch case statements

'win32' and 'linux' case statements yield the same results

stanier committed:
Replaced pipes with double pipes

(Why didn't the linter catch this?)

stanier committed:
Replaced double pipes with 'or'

stanier commented:
Passes unit tests now.

As a side note, this is why I generally dislike CoffeeScript (see a6d4b91 d9c68a4)

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+          @emit 'update-not-available'
+          return
+
+        console.log 'will update'
```
;o

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+class AutoUpdater
+  _.extend @prototype, EventEmitter.prototype
+
+  stdout = ChildProcess.execSync('cat /etc/issue').toString().replace(/\\./g, "")
```
I believe Atom uses `lsb_release` to determine the OS - could you do that instead?
https://github.com/atom/notifications/blob/4716ca24f74c685f6db85447e432e8a173d86a67/lib/user-utilities.coffee#L64-L78 (though it does seem a bit broken on Fedora)

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+  else if distros.fedora.test(stdout)
+    @pkg = 'atom.x86_64.rpm'
+    @pkgType = 'rpm'
+    @installCmd = 'dnf install -y'
```
Does this work on Fedora 21 and under?

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+      try
+        update = JSON.parse(stdout)
+
+        spawn 'atom', ['--version'], (error, stdout) ->
```
The command isn't necessarily `atom` anymore with the introduction of beta releases - `atom-beta` is valid now and will launch the beta version which also supports auto-updates.  You'll need to catch that.

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+    app.quit()
+
+  downloadUpdate: (callback) ->
+    spawn 'curl', ['-L', @updateUrl], (error, stdout) ->
```
Is there any chance at all that `curl` isn't installed?  If so we'll need to add it as a dependency.

50Wliu commented on src/browser/auto-updater-linux.coffee:
```diff
@@ -0,0 +1,115 @@
+  setFeedUrl: (@updateUrl) ->
+
+  quitAndInstall: ->
+    app.once 'will-quit', -> spawn 'sudo', [installCmd, @pkg, '&&',
```
`sudo` :(
I'm guessing that this can't be avoided though.

50Wliu commented:
The linux packages are only for 64-bit right now, but I'm fairly certain that 32-bit users can build it manually and run Atom (possibly pinning it to a release version and enabling updates).  If so, that should be addressed in the auto-updater so that it doesn't try installing 64-bit builds on 32-bit machines.

stanier commented:
... I'll just wait for the official repositories

lee-dohm commented:
Thanks @stanier for all your hard work on this. We talked this over and we're not comfortable with supporting a partial solution like this. We're still planning on having support for the official repositories. I don't have an ETA on when that will be accomplished right now though.

Thanks again for driving this!
